lib/pos_mempool.go

```diff
@@ -0,0 +1,490 @@
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem validating transaction sanity")
+	}
+
+	if err := dmp.latestBlockView.ValidateTransactionNonce(txn, dmp.latestBlockHeight); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Problem validating transaction nonce")
+	}
+
+	// Check transaction signature.
+	if _, err := dmp.latestBlockView.VerifySignature(txn, uint32(dmp.latestBlockHeight)); err != nil {
+		return errors.Wrapf(err, "PosMempool.AddTransaction: Signature validation failed")
+	}
+
+	return nil
+}
+
+func (dmp *PosMempool) addTransactionNoLock(txn *MempoolTx, persistToDb bool) error {
+	userPk := NewPublicKey(txn.Tx.PublicKey)
+	txnFee := txn.Tx.TxnFeeNanos
+
+	// Validate that the user has enough balance to cover the transaction fees.
+	spendableBalanceNanos, err := dmp.latestBlockView.GetSpendableDeSoBalanceNanosForPublicKey(userPk.ToBytes(),
+		uint32(dmp.latestBlockHeight))
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem getting spendable balance")
+	}
+	if err := dmp.ledger.CanIncreaseEntryWithLimit(*userPk, txnFee, spendableBalanceNanos); err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem checking balance increase for transaction with"+
+			"hash %v, fee %v", txn.Tx.Hash(), txnFee)
+	}
+
+	// If we get here, it means that the transaction's sender has enough balance to cover transaction fees. We can now
+	// add the transaction to mempool.
+	if err := dmp.txnRegister.AddTransaction(txn); err != nil {
+		return errors.Wrapf(err, "PosMempool.addTransactionNoLock: Problem adding txn to register")
+	}
+
+	// We update the reserved balance to include the newly added transaction's fee.
+	dmp.ledger.IncreaseEntry(*userPk, txnFee)
+
+	// Emit an event for the newly added transaction.
+	if persistToDb {
+		event := &MempoolEvent{
+			Txn:  txn,
+			Type: MempoolEventAdd,
+		}
+		dmp.persister.EnqueueEvent(event)
+	}
+
+	return nil
+}
+
+// loadPersistedTransactions fetches transactions from the persister's storage and adds the transactions to the mempool.
+// No lock is held and (persistToDb = false) flag is used when adding transactions internally.
+func (dmp *PosMempool) loadPersistedTransactions() error {
+	txns, err := dmp.persister.GetPersistedTransactions()
+	if err != nil {
+		return errors.Wrapf(err, "PosMempool.Start: Problem retrieving transactions from persister")
+	}
+	// We set the persistToDb flag to false so that persister doesn't try to save the transactions.
+	for _, txn := range txns {
+		if err := dmp.addTransactionNoLock(txn, false); err != nil {
+			glog.Errorf("PosMempool.Start: Problem adding transaction with hash (%v) from persister: %v",
+				txn.Hash, err)
+			continue
+		}
+		// If the transaction was successfully added, also include it in the nonce tracker.
+		pk := NewPublicKey(txn.Tx.PublicKey)
+		dmp.nonceTracker.AddTxnByPublicKeyNonce(*pk, *txn.Tx.TxnNonce, txn)
```

Comment:
If nonceTracker was a part of addTransactionNoLock, you could skip this second portion.